Cubical/HITs/Cylinder.agda

```diff
@@ -7,65 +7,263 @@ open import Cubical.Core.Everything
+  -- of type A ⊎ A → A into a pair of mappings:
+  --
+  --   A ⊎ A → Cylinder A ≃ A
+  --
+  -- include is the first part of the factorization.
+  include : A ⊎ A → Cylinder A
   include (Basics.inl x) = inl x
   include (Basics.inr x) = inr x
 
-  module _ {ℓ'} {B : Cylinder → Set ℓ'}
-                (f : (x : A) → B (inl x))
-                (g : (x : A) → B (inr x))
-                (p : ∀ x → PathP (λ i → B (cross x i)) (f x) (g x))
-    where
-    elimCyl : (c : Cylinder) → B c
-    elimCyl (inl x) = f x
-    elimCyl (inr x) = g x
-    elimCyl (cross x i) = p x i
+  -- The above inclusion is surjective
+  includeSurjective : ∀ c → ∥ Σ[ s ∈ A ⊎ A ] include s ≡ c ∥
+  includeSurjective (inl x) = ∣ Basics.inl x , refl ∣
+  includeSurjective (inr x) = ∣ Basics.inr x , refl ∣
+  includeSurjective (cross x i) =
+    squash
+      ∣ Basics.inl x , (λ j → cross x (i ∧ j)) ∣
+      ∣ Basics.inr x , (λ j → cross x (i ∨ ~ j)) ∣
+      i
+
+  elimCyl
+    : ∀{ℓ'} {B : Cylinder A → Set ℓ'}
+    → (f : (x : A) → B (inl x))
+    → (g : (x : A) → B (inr x))
+    → (p : ∀ x → PathP (λ i → B (cross x i)) (f x) (g x))
+    → (c : Cylinder A) → B c
+  elimCyl f _ _ (inl x) = f x
+  elimCyl _ g _ (inr x) = g x
+  elimCyl _ _ p (cross x i) = p x i
 
   private
-    out : Cylinder → A
+    out : Cylinder A → A
     out (inl x) = x
     out (inr x) = x
     out (cross x i) = x
 
-    inl-out : ∀ c → inl (out c) ≡ c
+    inl-out : (c : Cylinder A) → inl (out c) ≡ c
     inl-out (inl x) = refl
     inl-out (inr x) = cross x
     inl-out (cross x i) = λ j → cross x (i ∧ j)
 
-    out-inl : ∀ x → out (inl x) ≡ x
+    out-inl : ∀(x : A) → out (inl x) ≡ x
     out-inl x = refl
 
-  cylEquiv : Cylinder ≃ A
-  cylEquiv = isoToEquiv out inl out-inl inl-out
+  -- The second part of the factorization above.
+  CylinderA≃A : Cylinder A ≃ A
+  CylinderA≃A = isoToEquiv out inl out-inl inl-out
+
+  -- The cocylinder has a similar equivalence that is part
+  -- of factorizing the diagonal mapping.
+  private
+    inco : A → Cocylinder A
+    inco x _ = x
+
+    outco : Cocylinder A → A
+    outco f = f zero
+
+    A→CocylinderA→A : (x : A) → outco (inco x) ≡ x
+    A→CocylinderA→A x = refl
+
+    CocylinderA→A→CocylinderA : (c : Cocylinder A) → inco (outco c) ≡ c
+    CocylinderA→A→CocylinderA c j zero = c zero
+    CocylinderA→A→CocylinderA c j one = c (seg j)
+    CocylinderA→A→CocylinderA c j (seg i) = c (seg (j ∧ i))
+
+  A≃CocylinderA : A ≃ Cocylinder A
+  A≃CocylinderA =
+    isoToEquiv
+      inco outco
+      CocylinderA→A→CocylinderA
+      A→CocylinderA→A
+
+  project : Cocylinder A → A × A
+  project c = c zero , c one
+
+
+-- Since we can construct cylinders for every type, Cylinder actually
+-- constitutes a cylinder functor:
+--
+--   https://ncatlab.org/nlab/show/cylinder+functor
+--
+-- e₀ = inl
+-- e₁ = inr
+-- σ = out
+module Functorial where
+  private
+    variable
+      ℓa ℓb ℓc : Level
+      A : Set ℓa
+      B : Set ℓb
+      C : Set ℓc
```

Comment:
This can be made
```agda
    variable
      ℓ : Level
      A B C : Set ℓ
```

`A B C` will have different levels.

Reply:
By the way, is there documentation on the `variable` feature somewhere? I would not have guessed that it works this way.

Reply:
Same here, this behavior is very surprising to me

Reply:
Currently it's only documented in the CHANGELOG it seems:

https://github.com/agda/agda/blob/master/CHANGELOG.md